Implement the ability to provide a query with no filtering

akhtariev opened the pull request:
I have a use case where I need to get all actions without any filters applied. The current design does not allow for that. As a result, for all API functions that would take queries as parameters, I refactored out a query interface that has specific implementation for each API under `query` package. This is a breaking change.

- [x] allow for no filter passed by implementing a Query interface
- [x] refactor to reduce code duplication
- [x] add `gomega` module to use its matchers in unit tests 
- [x] unit test new functionality
- [x] all unit tests pass
- [x] minimize duplication in `newton.go`
- [x] unit test `Do`
- [x] implement CancelOrder (missing from API)

In the future, I'd like to suggest to incorporate a `ginkgo` BDD testing framework. It works well with gomega and really does simplify writing unit tests. We can also create some CI jobs to auto run unit tests that do not require Newton client and secret key.

@dhiaayachi please let me know what you think

dhiaayachi commented:
👋  @akhtariev, are you still interested in pursuing this PR? I can do another pass of reviews whenever it's ready.

akhtariev commented:
Hi @dhiaayachi ! Yes, I'll finish it up in the next couple of days. Sorry I've been busy lately

dhiaayachi commented:
No worries. Take your time and ping me when you have something ready for a review.

Thank you for contributing !!

akhtariev commented on newton.go:
```diff
@@ -329,21 +307,15 @@ func (n *Newton) HealthCheck() error {
 
 func (n *Newton) GetMinimumTradeAmount() (*GetMinTradeAmountsResp, error) {
-	res, err := n.doPublicQuery("/order/minimums", http.MethodGet, []Args{}, "")
+	res, err := n.doQuery("/order/minimums", http.MethodGet, []query.Parameter{}, true, "")
```
Implemented it similarly to `httpClient`'s `Do`

akhtariev commented on newton.go:
```diff
@@ -408,38 +348,15 @@ func (n *Newton) Balances(asset string) (*BalancesResp, error) {
-
-	res, err := n.doPrivateQuery("/actions", http.MethodGet, a, "")
+func (n *Newton) Actions(query query.Query) (*ActionsResp, error) {
+	res, err := n.doQuery("/actions", http.MethodGet, query.GetParameters(), false, "")
```
Not applicable anymore after `Do` refactor

akhtariev commented on newton.go:
```diff
@@ -451,44 +368,18 @@ func (n *Newton) Actions(actionType ActionType, limit int, offset int, startDate
-
-	res, err := n.doPrivateQuery("/order/history", http.MethodGet, a, "")
+func (n *Newton) OrderHistory(query query.Query) (*OrderHistoryResp, error) {
+	res, err := n.doQuery("/order/history", http.MethodGet, query.GetParameters(), false, "")
```
Not applicable anymore after `Do` refactor

akhtariev commented on newton.go:
```diff
@@ -311,16 +294,11 @@ func (n *Newton) GetSymbols(baseAsset, quoteAsset string) (*GetSymbolsResp, erro
 
 func (n *Newton) HealthCheck() error {
-	res, err := n.doPublicQuery("/symbols", http.MethodGet, []Args{}, "")
+	res, err := n.doQuery("/symbols", http.MethodGet, []query.Parameter{}, true, "")
```
fixed

akhtariev commented on newton.go:
```diff
@@ -497,37 +388,16 @@ func (n *Newton) OrdersHistory(limit int, offset int, startDate int64, endDate i
-
-	res, err := n.doPrivateQuery("/order/history", http.MethodGet, a, "")
+func (n *Newton) OpenOrders(query query.Query) (*OpenOrdersResp, error) {
+	res, err := n.doQuery("/order/history", http.MethodGet, query.GetParameters(), false, "")
```
Not applicable anymore after `Do` refactor

akhtariev commented on newton.go:
```diff
@@ -547,23 +417,16 @@ func (n *Newton) NewOrder(orderType string, timeInForce string, side string, sym
 	}
 
-	res, err := n.doPrivateQuery("/order/new", http.MethodPost, nil, string(b))
+	res, err := n.doQuery("/order/new", http.MethodPost, nil, false, string(b))
```
`NewOrder` is now using the same pattern

akhtariev commented:
Hi @dhiaayachi ! Hope your weekend is going well. I just finished addressing your comments. I ended up refactoring it ti make it look similar to `httpClient` `Do` right away. I also implemented `CancelOrder` endpoint.

In addition, I noticed that Newton now has a mock server so I switched every unit test to query that mock server. However, the mock server does not test authentication so I added another test for that that must be enabled by setting `TEST_AUTH` environment variable. I added more info on that in the `README`. 

Going forward, we can automate unit tests with GitHub Actions. Also, since the mock server simply hardcodes data in response, we can make more assertions on the response in the unit tests. Just to ensure that every response has the fields we expect it to have. However, we should probably do that in separate PRs.

Let me know what you think

dhiaayachi commented on newton.go:
```diff
@@ -160,415 +73,93 @@ func (n *Newton) sign(req *http.Request) error {
 
-	var r ActionsResp
-	err = json.Unmarshal(body, &r.Actions)
+	parsedResponse, err := n.parseResponse(res, query.GetResponse())
```
This is for future improvements, but I think that we can make `GetRepsonse` return a defined interface, that wrap `map[string]interface{}` and have methods that get fields to the enduser, like `GetAsString` or `GetAsFloat`...

akhtariev replied:
Created [issue](https://github.com/dhiaayachi/go-newton-co/issues/8)

dhiaayachi commented on newton.go:
```diff
@@ -160,415 +73,93 @@ func (n *Newton) sign(req *http.Request) error {
-	if res.StatusCode != http.StatusOK {
-		return nil, errors.New(fmt.Sprintf("request failed :: %d", res.StatusCode))
+	if toParseTo == nil {
+		return parsedResponse, nil
 	}
 
-	body, _ := ioutil.ReadAll(res.Body)
+	body, err := ioutil.ReadAll(res.Body)
```
I think we should read the body first and then decide if we return a nil value and error if a body is returned and `toParseTo` is `nil`, the reason is that if the body of a request is not read properly by the client it can reset the connexion and the body is never properly freed.

akhtariev replied:
Could you elaborate on this please? I'd assume it is safe since we close response's Body in defer. Why do we need to read the body in addition to closing to ensure that it is properly freed?

dhiaayachi replied:
there is a special behaviour in go http client that if you close the body without reading the buffer it will reset the underlying connexion and that could be a source of performance issues. [more details in here](https://stackoverflow.com/questions/18598780/is-resp-body-close-necessary-if-we-dont-read-anything-from-the-body)

akhtariev replied:
thanks for the explanation

dhiaayachi commented on query/actions.go:
```diff
@@ -0,0 +1,72 @@
+		params = append(params, Parameter{string(ActionTypeKey), string(a.ActionType)})
+	}
+
+	if a.Limit != int(ANY) {
```
the cast here is not needed

dhiaayachi commented on query/actions.go:
```diff
@@ -0,0 +1,72 @@
+		params = append(params, Parameter{string(Limit), strconv.Itoa(a.Limit)})
+	}
+
+	if a.Offset != int(ANY) {
```
same here for casting `ANY`

dhiaayachi commented on export_test.go:
```diff
@@ -0,0 +1,5 @@
+package newton
+
+const (
+	BaseURL = baseUrl
+)
```
I don't think this file is needed, we can use `New` in the auth test and for verifying the prod url, I prefer to have another const in the test package, so if someone change the prod url he need to change it in 2 places to make sure it's a willing change.

dhiaayachi commented on auth_test.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+TEST_AUTH=true CLIENT_ID=$1 CLIENT_SECRET=$2 go test
```
This is ok for the time being but I think I can make even the auth test run in automated environment, I will start a PR for this as soon as this is merged

akhtariev replied:
Sounds great. Would be curious to see how you make it work

dhiaayachi replied:
I have some ideas, I will let you know the details if my experiment succeed 😃

dhiaayachi reviewed (commented):
Thank you for this great contribution @akhtariev. This is really good!! My weekend was good and you made it even better with this contribution 😃.
I made a few comments, but nothing really big, this is now really close. Can you please also gofmt the code, I'm planing on adding a linting step into the automated CI.

akhtariev commented:
@dhiaayachi Thanks for another round of review! I addressed all comments. However, the reasoning behind [this one](https://github.com/dhiaayachi/go-newton-co/pull/7#discussion_r748962073) isn't exactly clear. Would you mind elaborating on that please just for me to learn

dhiaayachi reviewed (approved):
Great work @akhtariev !! 🎉 
I will work on the CI as soon as I can, also the small change about the `Do` func signature. After that I will crank the release version to 0.10, as this is a fairly big API change.
